pkg/apis/seedmanagement/types_managedseed.go

```diff
@@ -0,0 +1,182 @@
+// Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package seedmanagement
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	gardenletconfigv1alpha1 "github.com/gardener/gardener/pkg/gardenlet/apis/config/v1alpha1"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ManagedSeed represents a Shoot that is registered as Seed.
+type ManagedSeed struct {
+	metav1.TypeMeta
+	// Standard object metadata.
+	metav1.ObjectMeta
+	// Specification of the ManagedSeed.
+	Spec ManagedSeedSpec
+	// Most recently observed status of the ManagedSeed.
+	Status ManagedSeedStatus
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ManagedSeedList is a list of ManagedSeed objects.
+type ManagedSeedList struct {
+	metav1.TypeMeta
+	// Standard list object metadata.
+	metav1.ListMeta
+	// Items is the list of ManagedSeeds.
+	Items []ManagedSeed
+}
+
+// ManagedSeedSpec is the specification of a ManagedSeed.
+type ManagedSeedSpec struct {
+	// Shoot is the Shoot that will be registered as Seed.
+	Shoot *Shoot
```

Comment:
WDYT about not tying a ManagedSeed not to a Shoot, but to a Kubernetes cluster in general? 
With the `plant` Gardener already has the ability to specify any external cluster.
A `ManagedSeed` could then be used to set up a Seed on any cluster with a public endpoint. 

In addition, it might be useful to register the Garden cluster as a Seed cluster (Soil). 
At the moment only the configuration of the initial Gardenlet (& Gardenlets behind firewalls) have to be deployed manually via a config map. 
Have not thought this completely through yet, but this could enable the operator to manage the initial Gardenlets configuration using a `ManagedSeed`. This way, the Gardenlet could theoretically manage its own configuration via a `ManagedSeed`. 

For the initial Garden landscape setup:
 - Deploy an initial Gardenlet 
 - Create plant referencing the Garden cluster
 - Create  `ManagedSeed` referencing the plant
 - Gardenlet deploys the new Gardenlet based on the `ManagedSeed` and hereby overwriting its own configuration.

Reply:
I know this would have been something for the GEP phase, but I just thought about it now. Maybe I have missed it and it was already discussed previously.

Reply:
This sounds very complex, TBH, I would refrain from introducing this at this point. `Plant`s are not well established and were just contributed but never developed further. I'm not even sure if we want to keep them or rather drop their entire support. I'm not aware of any real-world usages.

Reply:
> Plants are not well established and were just contributed but never developed further. I'm not even sure if we want to keep them or rather drop their entire support. I'm not aware of any real-world usages.

To me, it is rather because we never attempted to build anything useful with it  + never even advertised in the dashboard etc. 
In this case, I see a clear benefit of using plants to bootstrap seeds on clusters not managed by Gardener. Also from a user perspective, I do not see why we would only allow Seeds to be managed if the cluster is a Shoot.

>  I would refrain from introducing this at this point

We do not need to include this now, I am only raising this to keep this in mind for the API design.
For instance, we could design it to be extensible at a later point:

```
type ManagedSeedSpec struct {
    TargetCluster *TargetCluster
}

type targetCluster struct {
      Shoot *Shoot
      Plant *Plant  (introduce later)
}

```

Reply:
> I do not see why we would only allow Seeds to be managed if the cluster is a Shoot.

Agreed, I guess this was also @mandelsoft's point in a discussion last year. We talked about this and decided that for now, because the scope is already quite large, we reflect this in the API but do not build more than the "managed seed is a shoot" use-case that exists today. In the future, it could be extended if necessary to also support other cluster types (e.g., "managed seed is a plant" or "managed seed is a GKE" or whatever).

The API as of now does not introduce a separate `TargetCluster` struct if that's your point. Personally, I don't see a need for it (the current design is similar to `.spec.volumes` in `Pod`s). Is there a benefit of wrapping this into this `TargetCluster` struct?

Reply:
I thought that this way we can cleanly add additional cluster references later - e.g if we would like to add Plants in the future.

Mhm not sure how the current design is similar to pods - the `.spec.volumes` in `Pod` has a `VolumeSource` with different kind of disks. Above I propose something similar with multiple different target clusters in one struct.

Reply:
Regarding the API design, `Shoot` is currently `+optional` and a pointer exactly because of this consideration, since as @rfranzke pointed out it was discussed and accepted:

```go
type ManagedSeedSpec struct {
    // +optional
    Shoot *Shoot
    // +optional (introduce later)
    Plant *Plant
    ...
}
```

I don't think we need a special `TargetCluster` type for this, having such alternative fields embedded in a larger structure instead of introducing a special one is common both in Go and in Kubernetes. We already do something similar with `SeedTemplate` and `Gardenlet`.

@danielfoehrKn In another comment you suggested making `Shoot` non-optional as this is currently the only supported used case. I tend to agree, it would simplify things for now and could always be changed later without being an incompatible change. @rfranzke WDYT?

Reply:
> I don't think we need a special TargetCluster type for this

@stoyanr okay, sure

> In another comment you suggested making Shoot non-optional as this is currently the only supported used case. I tend to agree, it would simplify things for now and could always be changed later without being an incompatible change. 

Sounds good

Reply:
We decide to make `Shoot` non-optional and not a pointer, since it's the only supported use case. The considerations for future extension mentioned above still apply.